File: SmartPresenceFrontend/src/app/core/services/etudiant.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { API, extraireDonnees } from '../api';
import { ApiResponse } from '../models/api.model';
import { Classe } from '../models/classe.model';
import { CompteEtudiant, Etudiant, EtudiantRequest } from '../models/etudiant.model';

/** Référentiel des étudiants, enrôlement biométrique et accès mobile. */
@Injectable({ providedIn: 'root' })
export class EtudiantService {
  private readonly http = inject(HttpClient);
  private readonly base = `${API}/etudiants`;

  lister(classeId?: number | null): Observable<Etudiant[]> {
    let params = new HttpParams();
    if (classeId) params = params.set('classeId', String(classeId));
    return extraireDonnees(this.http.get<ApiResponse<Etudiant[]>>(this.base, { params }));
  }

  creer(etudiant: EtudiantRequest): Observable<Etudiant> {
    return extraireDonnees(this.http.post<ApiResponse<Etudiant>>(this.base, etudiant));
  }

  modifier(id: string, etudiant: EtudiantRequest): Observable<Etudiant> {
    return extraireDonnees(this.http.put<ApiResponse<Etudiant>>(`${this.base}/${id}`, etudiant));
  }

  changerActivite(id: string, actif: boolean): Observable<void> {
    const params = new HttpParams().set('value', String(actif));
    return extraireDonnees(
      this.http.patch<ApiResponse<void>>(`${this.base}/${id}/active`, null, { params }),
    );
  }

  /** Associe la référence logique de l'empreinte capturée sur le lecteur. */
  enroler(id: string, biometricId: string): Observable<Etudiant> {
    return extraireDonnees(
      this.http.post<ApiResponse<Etudiant>>(`${this.base}/${id}/enrolement`, { biometricId }),
    );
  }

  revoquerEnrolement(id: string): Observable<Etudiant> {
    return extraireDonnees(this.http.delete<ApiResponse<Etudiant>>(`${this.base}/${id}/enrolement`));
  }

  /** Ouvre l'accès mobile et renvoie les identifiants — affichés une seule fois. */
  ouvrirCompte(id: string): Observable<CompteEtudiant> {
    return extraireDonnees(
      this.http.post<ApiResponse<CompteEtudiant>>(`${this.base}/${id}/compte`, null),
    );
  }

  fermerCompte(id: string): Observable<Etudiant> {
    return extraireDonnees(this.http.delete<ApiResponse<Etudiant>>(`${this.base}/${id}/compte`));
  }

  /** Classes disponibles pour l'affectation. */
  listerClasses(): Observable<Classe[]> {
    return extraireDonnees(this.http.get<ApiResponse<Classe[]>>(`${API}/classes`));
  }
}

File: BackendSmartPresence/src/main/java/com/smartpresence/service/EtudiantService.java
package com.smartpresence.service;

import com.smartpresence.dto.request.EtudiantRequest;
import com.smartpresence.dto.response.CompteEtudiantResponse;
import com.smartpresence.dto.response.EtudiantResponse;
import java.util.List;
import java.util.UUID;

public interface EtudiantService {

    List<EtudiantResponse> findAll(Long classeId);

    EtudiantResponse findById(UUID id);

    EtudiantResponse create(EtudiantRequest request);

    EtudiantResponse update(UUID id, EtudiantRequest request);

    void setActive(UUID id, boolean actif);

    /**
     * Associe à l'étudiant la référence logique de son empreinte.
     *
     * @param id          étudiant concerné
     * @param biometricId référence produite lors de la capture sur le lecteur
     */
    EtudiantResponse enroler(UUID id, String biometricId);

    /** Révoque l'enrôlement biométrique, en conservant les présences déjà relevées. */
    EtudiantResponse revoquerEnrolement(UUID id);

    /**
     * Ouvre un accès mobile à l'étudiant : crée son compte, lui attribue le rôle
     * {@code ETUDIANT} et le rattache à sa fiche.
     *
     * <p>Le mot de passe est tiré au hasard et retourné <b>une seule fois</b> : il est
     * haché avant enregistrement et le serveur ne peut plus le restituer.</p>
     *
     * @param id étudiant concerné
     * @return les identifiants à remettre à l'étudiant
     */
    CompteEtudiantResponse ouvrirCompte(UUID id);

    /**
     * Ferme l'accès mobile d'un étudiant et supprime son compte.
     *
     * <p>Les présences déjà relevées sont conservées : elles appartiennent au dossier
     * de scolarité, pas au compte de consultation.</p>
     *
     * @param id étudiant concerné
     */
    EtudiantResponse fermerCompte(UUID id);
}
